staging/src/k8s.io/apiserver/pkg/admission/metrics.go

```diff
@@ -0,0 +1,151 @@
+
+	// Built-in admission controller metrics. Each admission controller is identified by name.
+	controller := newMetricSet("controller",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "rejected"},
+		"Admission controller %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+
+	// External admission webhook metrics. Each webhook is identified by name.
+	externalWebhook := newMetricSet("external_webhook",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "rejected"},
+		"External admission webhook %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+
+	step.mustRegister()
+	controller.mustRegister()
+	externalWebhook.mustRegister()
+	return &AdmissionMetrics{step: step, controller: controller, externalWebhook: externalWebhook}
+}
+
+func (m *AdmissionMetrics) reset() {
+	m.step.reset()
+	m.controller.reset()
+	m.externalWebhook.reset()
+}
+
+// ObserveAdmissionStep records admission related metrics for a admission step, identified by step type.
+func (m *AdmissionMetrics) ObserveAdmissionStep(elapsed time.Duration, rejected bool, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	m.step.observe(elapsed, stepType, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+// ObserveAdmissionController records admission related metrics for a built-in admission controller, identified by it's plugin handler name.
+func (m *AdmissionMetrics) ObserveAdmissionController(elapsed time.Duration, rejected bool, handler NamedHandler, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	m.controller.observe(elapsed, handler.Name(), stepType, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+// ObserveExternalWebhook records admission related metrics for a external admission webhook.
+func (m *AdmissionMetrics) ObserveExternalWebhook(elapsed time.Duration, rejected bool, hook *v1alpha1.Webhook, attr Attributes) {
+	t := "validating" // TODO: pass in type (validating|mutating) once mutating webhook functionality has been implemented
+	gvr := attr.GetResource()
+	m.externalWebhook.observe(elapsed, hook.Name, t, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), strconv.FormatBool(rejected))
+}
+
+type metricSet struct {
+	latencies        *prometheus.HistogramVec
+	latenciesSummary *prometheus.SummaryVec
+}
+
+func newMetricSet(name string, labels []string, helpTemplate string) *metricSet {
+	return &metricSet{
+		latencies: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      fmt.Sprintf("%s_latencies", name),
```

Comment:
Should be `fmt.Sprintf("%s_request_latencies_microseconds")`, see https://prometheus.io/docs/practices/naming/. It's currently super unclear what `admission_step_latencies` mean, maybe I'm even wrong and those are not requests which only proves my point :)

Additionally we have bunch of metrics with different units, so adding it to the metric name is also a good idea

Reply:
We originally had the  unit on the metric names and subsequently removed them. This was discussed in the design doc (https://docs.google.com/document/d/1rDRrC5MNI2An_FeYx8HAsmlQv1WThh9IBb3_shCTbJA/edit#bookmark=id.z2fplna0fwhs). The decision was to favor consistency with existing kube-apiserver metric names already in use.

Each of metrics are on a "admission" operation.

Reply:
TBH, I don't see any actual discussion and there are several metrics in this repo which expose the metrics in a correct way already:

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/metrics/prometheus/prometheus.go
https://github.com/kubernetes/kubernetes/blob/master/pkg/util/reflector/prometheus/prometheus.go

I don't think being consistent with some subset of metrics is a good enough reason to ditch the guidelines. FWIW, it's a good way to gradually enforce the standard

Reply:
I'm happy to apply guidelines if we're aspiring toward a specific set of them. Is the guidance from sig-instrumentation to follow the prometheus naming guidelines? Any supplemental guidelines we following?  My understanding of the prometheus naming guide is that I should have something like: "%_<event>_latencies_seconds", does that sound right?

Reply:
I've gone with _seconds_ based on my understanding of prometheus naming conventions.

Reply:
Thanks a lot! We're always happy to help, feel free to ping me or @piosz :)

> Is the guidance from sig-instrumentation to follow the prometheus naming guidelines?

Yes

> Any supplemental guidelines we following? 

Not yet. As with the logging verbosity levels, this is a problem nobody had time to work on :(

> My understanding of the prometheus naming guide is that I should have something like: "%__latencies_seconds", does that sound right?

SGTM, from working with other metrics, that seems like a great idea, to use seconds instead (nano|micro|milli)seconds